src/Compiler/TypedTree/CompilerGlobalState.fs

```diff
@@ -15,62 +18,31 @@ open FSharp.Compiler.Text
 /// are used to host multiple concurrent instances of compilation.
-type NiceNameGenerator() =
-
-    let lockObj = obj()
-    let basicNameCounts = Dictionary<string, int>(100)
-
+type NiceNameGenerator() =    
+    let basicNameCounts = ConcurrentDictionary<string,Ref<int>>(max Environment.ProcessorCount 1, 127)
+   
     member _.FreshCompilerGeneratedName (name, m: range) =
-      lock lockObj (fun () ->
         let basicName = GetBasicNameOfPossibleCompilerGeneratedName name
-        let n =
-            match basicNameCounts.TryGetValue basicName with
-            | true, count -> count
-            | _ -> 0
-        let nm = CompilerGeneratedNameSuffix basicName (string m.StartLine + (match n with 0 -> "" | n -> "-" + string n))
-        basicNameCounts[basicName] <- n + 1
-        nm)
-
-    member _.Reset () =
-      lock lockObj (fun () ->
-        basicNameCounts.Clear()
-      )
+        let countCell = basicNameCounts.GetOrAdd(basicName,fun k -> ref 0)
+        let count = Interlocked.Increment(countCell)
+        
+        CompilerGeneratedNameSuffix basicName (string m.StartLine + (match (count-1) with 0 -> "" | n -> "-" + string n))
 
 /// Generates compiler-generated names marked up with a source code location, but if given the same unique value then
 /// return precisely the same name. Each name generated also includes the StartLine number of the range passed in
 /// at the point of first generation.
 ///
 /// This type may be accessed concurrently, though in practice it is only used from the compilation thread.
 /// It is made concurrency-safe since a global instance of the type is allocated in tast.fs.
-type StableNiceNameGenerator() =
+type StableNiceNameGenerator() = 
 
-    let lockObj = obj()
-
-    let names = Dictionary<string * int64, string>(100)
-    let basicNameCounts = Dictionary<string, int>(100)
+    let niceNames = ConcurrentDictionary<string * int64, string>(max Environment.ProcessorCount 1, 127)
+    let innerGenerator = new NiceNameGenerator()
 
     member x.GetUniqueCompilerGeneratedName (name, m: range, uniq) =
-        lock lockObj (fun () ->
-            let basicName = GetBasicNameOfPossibleCompilerGeneratedName name
-            let key = basicName, uniq
-            match names.TryGetValue key with
-            | true, nm -> nm
-            | _ ->
-                let n =
-                    match basicNameCounts.TryGetValue basicName with
-                    | true, c -> c
-                    | _ -> 0
-                let nm = CompilerGeneratedNameSuffix basicName (string m.StartLine + (match n with 0 -> "" | n -> "-" + string n))
-                names[key] <- nm
-                basicNameCounts[basicName] <- n + 1
-                nm
-        )
-
-    member x.Reset () =
-      lock lockObj (fun () ->
-        basicNameCounts.Clear()
-        names.Clear()
-      )
+        let basicName = GetBasicNameOfPossibleCompilerGeneratedName name
```

Comment:
The code of StableNiceNameGenerator and NiceNameGenerator was very much duplicate.
Changing so that StableNiceNameGenerator  builds on top of NiceNameGenerator .